airflow/providers/google/cloud/hooks/vertex_ai/custom_job.py

```diff
@@ -2178,3 +3300,239 @@ def list_custom_jobs(
+            timeout=timeout,
+            metadata=metadata,
+        )
+        return result
+
+    async def get_training_pipeline(
+        self,
+        project_id: str,
+        location: str,
+        pipeline_id: str,
+        retry: AsyncRetry | _MethodDefault = DEFAULT,
+        timeout: float | _MethodDefault | None = DEFAULT,
+        metadata: Sequence[tuple[str, str]] = (),
+        client: PipelineServiceAsyncClient | None = None,
+    ) -> types.TrainingPipeline:
+        """
+        Get a TrainingPipeline proto message from PipelineServiceAsyncClient.
+
+        :param project_id: Required. The ID of the Google Cloud project that the service belongs to.
+        :param location: Required. The ID of the Google Cloud region that the service belongs to.
+        :param pipeline_id: Required. The ID of the PipelineJob resource.
+        :param retry: Designation of what errors, if any, should be retried.
+        :param timeout: The timeout for this request.
+        :param metadata: Strings which should be sent along with the request as metadata.
+        """
+        if not client:
+            client = await self.get_pipeline_service_client(region=location)
+        pipeline_name = client.training_pipeline_path(
+            project=project_id,
+            location=location,
+            training_pipeline=pipeline_id,
+        )
+        response: types.TrainingPipeline = await client.get_training_pipeline(
+            request={"name": pipeline_name},
+            retry=retry,
+            timeout=timeout,
+            metadata=metadata,
+        )
+        return response
+
+    async def wait_for_custom_job(
+        self,
+        project_id: str,
+        location: str,
+        job_id: str,
+        retry: AsyncRetry | _MethodDefault = DEFAULT,
+        timeout: float | None = None,
+        metadata: Sequence[tuple[str, str]] = (),
+        poll_interval: int = 10,
+    ) -> types.CustomJob:
+        client = await self.get_job_service_client(region=location)
+        while True:
+            try:
+                self.log.info("Requesting a custom job with id %s", job_id)
+                job: types.CustomJob = await self.get_custom_job(
+                    project_id=project_id,
+                    location=location,
+                    job_id=job_id,
+                    retry=retry,
+                    timeout=timeout,
+                    metadata=metadata,
+                    client=client,
+                )
+            except Exception as ex:
+                self.log.exception("Exception occurred while requesting job %s", job_id)
+                raise AirflowException(ex)
+            self.log.info("Status of the custom job %s is %s", job.name, job.state.name)
+            if job.state in self.JOB_COMPLETE_STATES:
+                return job
+            self.log.info("Sleeping for %s seconds.", poll_interval)
+            await asyncio.sleep(poll_interval)
+
+    async def wait_for_training_pipeline(
```

Comment:
I feel the structure of this method is similar to wait_for_custom_job. Not sure whether we can extract the common part of them. Not a blocker of this PR though

Reply:
Indeed, the structure is deliberately similar. We tried to follow the same pattern for Vertex AI trigger and hook methods. 
The only difference is that ` wait_for_custom_job()` and `wait_for_training_pipeline()` use different clients and client methods.